File: android-app/app/src/main/java/com/bhan796/anagramarena/audio/SoundManager.kt
package com.bhan796.anagramarena.audio

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.PI
import kotlin.math.sin

private enum class MusicMode {
    NONE,
    MENU,
    MATCH
}

private enum class Waveform {
    SINE,
    SQUARE,
    SAW
}

private data class ToneStep(
    val frequencyHz: Double,
    val beats: Double,
    val waveform: Waveform
)

object SoundManager {
    private const val SAMPLE_RATE = 22_050
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private var soundEnabled = true
    private var masterMuted = false
    private var musicEnabled = true
    private var uiSfxEnabled = true
    private var gameSfxEnabled = true

    private var musicVolume = 0.5f
    private var uiSfxVolume = 0.8f
    private var gameSfxVolume = 0.85f

    private var activeMusicMode = MusicMode.NONE
    private var musicJob: Job? = null

    private fun canPlayUiSfx(): Boolean = soundEnabled && !masterMuted && uiSfxEnabled && uiSfxVolume > 0f
    private fun canPlayGameSfx(): Boolean = soundEnabled && !masterMuted && gameSfxEnabled && gameSfxVolume > 0f
    private fun canPlayMusic(): Boolean = !masterMuted && musicEnabled && musicVolume > 0f

    private fun playTone(frequencyHz: Double, durationMs: Int, volume: Float, waveform: Waveform = Waveform.SQUARE) {
        val level = volume.coerceIn(0f, 1f)
        if (level <= 0f) return
        scope.launch {
            playBuffer(generateBuffer(frequencyHz, durationMs, level, waveform))
        }
    }

    private fun playChord(tones: List<Double>, durationMs: Int, volume: Float, waveform: Waveform = Waveform.SQUARE) {
        val level = volume.coerceIn(0f, 1f)
        if (level <= 0f || tones.isEmpty()) return
        scope.launch {
            playBuffer(generateMixedBuffer(tones, durationMs, level, waveform))
        }
    }

    private fun playSequence(steps: List<Pair<Double, Int>>, volume: Float, waveform: Waveform = Waveform.SQUARE) {
        val level = volume.coerceIn(0f, 1f)
        if (level <= 0f || steps.isEmpty()) return
        scope.launch {
            for ((freq, duration) in steps) {
                playBuffer(generateBuffer(freq, duration, level, waveform))
                if (!isActive) break
            }
        }
    }

    private fun generateBuffer(frequencyHz: Double, durationMs: Int, volume: Float, waveform: Waveform): ShortArray {
        val sampleCount = ((SAMPLE_RATE * durationMs) / 1000.0).toInt().coerceAtLeast(1)
        val attackSamples = (sampleCount * 0.07).toInt().coerceAtLeast(1)
        val releaseSamples = (sampleCount * 0.15).toInt().coerceAtLeast(1)
        val data = ShortArray(sampleCount)

        for (i in 0 until sampleCount) {
            val phase = (2.0 * PI * frequencyHz * i.toDouble()) / SAMPLE_RATE.toDouble()
            val raw = when (waveform) {
                Waveform.SINE -> sin(phase)
                Waveform.SQUARE -> if (sin(phase) >= 0.0) 1.0 else -1.0
                Waveform.SAW -> ((i.toDouble() * frequencyHz / SAMPLE_RATE.toDouble()) % 1.0) * 2.0 - 1.0
            }

            val envelope = when {
                i < attackSamples -> i.toDouble() / attackSamples.toDouble()
                i > sampleCount - releaseSamples -> (sampleCount - i).toDouble() / releaseSamples.toDouble()
                else -> 1.0
            }.coerceIn(0.0, 1.0)

            val value = raw * envelope * volume
            data[i] = (value * Short.MAX_VALUE).toInt().coerceIn(Short.MIN_VALUE.toInt(), Short.MAX_VALUE.toInt()).toShort()
        }

        return data
    }

    private fun generateMixedBuffer(tones: List<Double>, durationMs: Int, volume: Float, waveform: Waveform): ShortArray {
        val sampleCount = ((SAMPLE_RATE * durationMs) / 1000.0).toInt().coerceAtLeast(1)
        val attackSamples = (sampleCount * 0.07).toInt().coerceAtLeast(1)
        val releaseSamples = (sampleCount * 0.15).toInt().coerceAtLeast(1)
        val data = ShortArray(sampleCount)
        val toneCount = tones.size.coerceAtLeast(1)

        for (i in 0 until sampleCount) {
            var mixed = 0.0
            for (frequencyHz in tones) {
                val phase = (2.0 * PI * frequencyHz * i.toDouble()) / SAMPLE_RATE.toDouble()
                val raw = when (waveform) {
                    Waveform.SINE -> sin(phase)
                    Waveform.SQUARE -> if (sin(phase) >= 0.0) 1.0 else -1.0
                    Waveform.SAW -> ((i.toDouble() * frequencyHz / SAMPLE_RATE.toDouble()) % 1.0) * 2.0 - 1.0
                }
                mixed += raw / toneCount.toDouble()
            }

            val envelope = when {
                i < attackSamples -> i.toDouble() / attackSamples.toDouble()
                i > sampleCount - releaseSamples -> (sampleCount - i).toDouble() / releaseSamples.toDouble()
                else -> 1.0
            }.coerceIn(0.0, 1.0)

            val value = mixed * envelope * volume
            data[i] = (value * Short.MAX_VALUE).toInt().coerceIn(Short.MIN_VALUE.toInt(), Short.MAX_VALUE.toInt()).toShort()
        }

        return data
    }

    private fun playBuffer(buffer: ShortArray) {
        val audioTrack = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_GAME)
                    .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(SAMPLE_RATE)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build()
            )
            .setTransferMode(AudioTrack.MODE_STATIC)
            .setBufferSizeInBytes(buffer.size * 2)
            .build()

        try {
            audioTrack.write(buffer, 0, buffer.size, AudioTrack.WRITE_BLOCKING)
            audioTrack.play()
            while (audioTrack.playState == AudioTrack.PLAYSTATE_PLAYING && audioTrack.playbackHeadPosition < buffer.size) {
                Thread.sleep(4)
            }
        } finally {
            audioTrack.release()
        }
    }

    private fun startLoop(mode: MusicMode, bpm: Int, pattern: List<ToneStep>) {
        musicJob?.cancel()
        musicJob = scope.launch {
            val beatMs = 60_000.0 / bpm.toDouble()
            while (isActive && activeMusicMode == mode && canPlayMusic()) {
                for (step in pattern) {
                    if (!isActive || activeMusicMode != mode || !canPlayMusic()) break
                    val duration = (beatMs * step.beats).toInt().coerceAtLeast(30)
                    playBuffer(generateBuffer(step.frequencyHz, duration, (musicVolume * 0.55f).coerceIn(0f, 1f), step.waveform))
                }
            }
        }
    }

    private fun updateMusicState(targetMode: MusicMode = activeMusicMode) {
        if (!canPlayMusic()) {
            stopMusic()
            return
        }

        when (targetMode) {
            MusicMode.NONE -> stopMusic()
            MusicMode.MENU -> startMenuMusic(forceRestart = true)
            MusicMode.MATCH -> startMatchMusic(forceRestart = true)
        }
    }

    fun setSoundEnabled(enabled: Boolean) {
        soundEnabled = enabled
    }

    fun setMasterMuted(muted: Boolean) {
        masterMuted = muted
        if (muted) {
            stopMusic()
        } else {
            updateMusicState()
        }
    }

    fun setMusicEnabled(enabled: Boolean) {
        musicEnabled = enabled
        if (!enabled) {
            stopMusic()
        } else {
            updateMusicState()
        }
    }

    fun setUiSfxEnabled(enabled: Boolean) {
        uiSfxEnabled = enabled
    }

    fun setGameSfxEnabled(enabled: Boolean) {
        gameSfxEnabled = enabled
    }

    fun setMusicVolume(volume: Float) {
        musicVolume = volume.coerceIn(0f, 1f)
        updateMusicState()
    }

    fun setUiSfxVolume(volume: Float) {
        uiSfxVolume = volume.coerceIn(0f, 1f)
    }

    fun setGameSfxVolume(volume: Float) {
        gameSfxVolume = volume.coerceIn(0f, 1f)
    }

    fun playClick() {
        if (!canPlayUiSfx()) return
        playTone(1046.50, 52, uiSfxVolume, Waveform.SQUARE)
    }

    fun playTilePlace() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(392.0 to 90, 329.63 to 90), gameSfxVolume, Waveform.SQUARE)
    }

    fun playWordSubmit() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(523.25 to 90, 659.25 to 90, 783.99 to 90), gameSfxVolume, Waveform.SINE)
    }

    fun playWordValid() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(523.25 to 80, 659.25 to 80, 783.99 to 80, 1046.50 to 80), gameSfxVolume, Waveform.SINE)
    }

    fun playWordInvalid() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(220.0 to 140, 174.61 to 120), gameSfxVolume, Waveform.SAW)
    }

    fun playTimerTick() {
        if (!canPlayGameSfx()) return
        playTone(1046.50, 42, gameSfxVolume * 0.8f, Waveform.SINE)
    }

    fun playTimerUrgent() {
        if (!canPlayGameSfx()) return
        playTone(1318.51, 70, gameSfxVolume, Waveform.SINE)
    }

    fun playMatchFound() {
        if (!canPlayUiSfx()) return
        playSequence(listOf(523.25 to 90, 659.25 to 90, 783.99 to 90, 1046.50 to 90), uiSfxVolume, Waveform.SQUARE)
    }

    fun playCountdownBeep() {
        if (!canPlayUiSfx()) return
        playTone(880.0, 120, uiSfxVolume, Waveform.SINE)
    }

    fun playCountdownGo() {
        if (!canPlayUiSfx()) return
        playChord(listOf(880.0, 1174.66, 1567.98), 260, uiSfxVolume, Waveform.SQUARE)
    }

    fun playRoundResult() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(783.99 to 100, 1046.50 to 100), gameSfxVolume, Waveform.SINE)
    }

    fun playWin() {
        if (!canPlayGameSfx()) return
        playSequence(
            listOf(523.25 to 90, 659.25 to 90, 783.99 to 90, 1046.50 to 90, 1318.51 to 90),
            gameSfxVolume,
            Waveform.SQUARE
        )
    }

    fun playLose() {
        if (!canPlayGameSfx()) return
        playSequence(listOf(392.0 to 120, 349.23 to 120, 311.13 to 120, 261.63 to 140), gameSfxVolume, Waveform.SAW)
    }

    fun startMenuMusic(forceRestart: Boolean = false) {
        if (!canPlayMusic()) return
        if (!forceRestart && activeMusicMode == MusicMode.MENU && musicJob?.isActive == true) return

        activeMusicMode = MusicMode.MENU
        startLoop(
            mode = MusicMode.MENU,
            bpm = 120,
            pattern = listOf(
                ToneStep(261.63, 0.5, Waveform.SQUARE),
                ToneStep(329.63, 0.5, Waveform.SQUARE),
                ToneStep(392.0, 0.5, Waveform.SQUARE),
                ToneStep(523.25, 0.5, Waveform.SQUARE),
                ToneStep(329.63, 0.5, Waveform.SQUARE),
                ToneStep(392.0, 0.5, Waveform.SQUARE),
                ToneStep(293.66, 0.5, Waveform.SQUARE),
                ToneStep(349.23, 0.5, Waveform.SQUARE)
            )
        )
    }

    fun startMatchMusic(forceRestart: Boolean = false) {
        if (!canPlayMusic()) return
        if (!forceRestart && activeMusicMode == MusicMode.MATCH && musicJob?.isActive == true) return

        activeMusicMode = MusicMode.MATCH
        startLoop(
            mode = MusicMode.MATCH,
            bpm = 140,
            pattern = listOf(
                ToneStep(220.0, 0.5, Waveform.SQUARE),
                ToneStep(261.63, 0.5, Waveform.SQUARE),
                ToneStep(329.63, 0.5, Waveform.SQUARE),
                ToneStep(440.0, 0.5, Waveform.SQUARE),
                ToneStep(392.0, 0.5, Waveform.SQUARE),
                ToneStep(329.63, 0.5, Waveform.SQUARE),
                ToneStep(261.63, 0.5, Waveform.SQUARE),
                ToneStep(220.0, 0.5, Waveform.SQUARE)
            )
        )
    }

    fun stopMusic() {
        musicJob?.cancel()
        musicJob = null
        activeMusicMode = MusicMode.NONE
    }
}

File: web-app/src/sound/SoundManager.ts
import * as Tone from "tone";

// -- Internal state ------------------------------------------------------------
let _soundEnabled = true;
let _masterMuted = false;

let _musicEnabled = true;
let _uiSfxEnabled = true;
let _gameSfxEnabled = true;

let _musicVolume = 0.5;
let _uiSfxVolume = 0.8;
let _gameSfxVolume = 0.85;

let _musicLoop: Tone.Part | null = null;
let _musicSynth: Tone.PolySynth | null = null;
let _musicReverb: Tone.Reverb | null = null;
let _started = false;

function clamp01(value: number): number {
  if (Number.isNaN(value)) return 0;
  return Math.max(0, Math.min(1, value));
}

function volumeToDb(volume: number): number {
  if (volume <= 0) return -72;
  return Tone.gainToDb(volume);
}

function canPlayUiSfx() {
  return _soundEnabled && !_masterMuted && _uiSfxEnabled && _uiSfxVolume > 0;
}

function canPlayGameSfx() {
  return _soundEnabled && !_masterMuted && _gameSfxEnabled && _gameSfxVolume > 0;
}

function canPlayMusic() {
  return !_masterMuted && _musicEnabled && _musicVolume > 0;
}

async function ensureStarted() {
  if (!_started) {
    await Tone.start();
    _started = true;
  }
}

// -- Settings ------------------------------------------------------------------
export function setSoundEnabled(enabled: boolean) {
  _soundEnabled = enabled;
}

export function setMasterMuted(muted: boolean) {
  _masterMuted = muted;
  if (muted) {
    stopMusic();
  }
}

export function setMusicEnabled(enabled: boolean) {
  _musicEnabled = enabled;
  if (!enabled) {
    stopMusic();
  }
}

export function setUiSfxEnabled(enabled: boolean) {
  _uiSfxEnabled = enabled;
}

export function setGameSfxEnabled(enabled: boolean) {
  _gameSfxEnabled = enabled;
}

export function setMusicVolume(volume: number) {
  _musicVolume = clamp01(volume);
  if (_musicSynth) {
    _musicSynth.volume.value = -18 + volumeToDb(_musicVolume);
  }
}

export function setUiSfxVolume(volume: number) {
  _uiSfxVolume = clamp01(volume);
}

export function setGameSfxVolume(volume: number) {
  _gameSfxVolume = clamp01(volume);
}

// -- Short SFX (fire-and-forget) -----------------------------------------------

export async function playClick() {
  if (!canPlayUiSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "square" },
    envelope: { attack: 0.001, decay: 0.04, sustain: 0, release: 0.01 },
    volume: -6 + volumeToDb(_uiSfxVolume)
  }).toDestination();
  synth.triggerAttackRelease("C6", "32n");
  window.setTimeout(() => synth.dispose(), 300);
}

export async function playTilePlace() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "square" },
    envelope: { attack: 0.001, decay: 0.07, sustain: 0, release: 0.01 },
    volume: -8 + volumeToDb(_gameSfxVolume)
  }).toDestination();
  synth.triggerAttackRelease("G4", "16n");
  window.setTimeout(() => {
    synth.triggerAttackRelease("E4", "16n");
  }, 80);
  window.setTimeout(() => synth.dispose(), 500);
}

export async function playWordSubmit() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "sine" },
    envelope: { attack: 0.01, decay: 0.1, sustain: 0.1, release: 0.2 },
    volume: -10 + volumeToDb(_gameSfxVolume)
  });
  const now = Tone.now();
  synth.triggerAttackRelease("C5", "8n", now);
  synth.triggerAttackRelease("E5", "8n", now + 0.08);
  synth.triggerAttackRelease("G5", "8n", now + 0.16);
  window.setTimeout(() => synth.dispose(), 1000);
}

export async function playWordValid() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "sine" },
    envelope: { attack: 0.01, decay: 0.12, sustain: 0.1, release: 0.3 },
    volume: -8 + volumeToDb(_gameSfxVolume)
  });
  const now = Tone.now();
  ["C5", "E5", "G5", "C6"].forEach((note, index) => {
    synth.triggerAttackRelease(note, "8n", now + index * 0.1);
  });
  window.setTimeout(() => synth.dispose(), 1500);
}

export async function playWordInvalid() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "square" },
    envelope: { attack: 0.01, decay: 0.2, sustain: 0, release: 0.05 },
    volume: -10 + volumeToDb(_gameSfxVolume)
  }).toDestination();
  const now = Tone.now();
  synth.triggerAttackRelease("A3", "8n", now);
  synth.triggerAttackRelease("F3", "8n", now + 0.15);
  window.setTimeout(() => synth.dispose(), 800);
}

export async function playTimerTick() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "sine" },
    envelope: { attack: 0.001, decay: 0.04, sustain: 0, release: 0.01 },
    volume: -12 + volumeToDb(_gameSfxVolume)
  }).toDestination();
  synth.triggerAttackRelease("C6", "32n");
  window.setTimeout(() => synth.dispose(), 300);
}

export async function playTimerUrgent() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "sine" },
    envelope: { attack: 0.001, decay: 0.06, sustain: 0, release: 0.01 },
    volume: -8 + volumeToDb(_gameSfxVolume)
  }).toDestination();
  synth.triggerAttackRelease("E6", "16n");
  window.setTimeout(() => synth.dispose(), 300);
}

export async function playMatchFound() {
  if (!canPlayUiSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "square" },
    envelope: { attack: 0.01, decay: 0.15, sustain: 0.1, release: 0.4 },
    volume: -4 + volumeToDb(_uiSfxVolume)
  });
  const now = Tone.now();
  ["C5", "E5", "G5", "C6", "E6"].forEach((note, index) => {
    synth.triggerAttackRelease(note, "8n", now + index * 0.12);
  });
  window.setTimeout(() => synth.dispose(), 2000);
}

export async function playLogoAssemble(): Promise<void> {
  if (!canPlayUiSfx()) return;
  await ensureStarted();

  const now = Tone.now();

  const noise = new Tone.Noise("pink").start(now).stop(now + 1.3);
  const filter = new Tone.Filter({ type: "bandpass", frequency: 120, Q: 0.8 });
  const gain = new Tone.Gain(0);
  noise.connect(filter);
  filter.connect(gain);
  gain.toDestination();

  gain.gain.setValueAtTime(0, now);
  gain.gain.linearRampToValueAtTime(0.18 * _uiSfxVolume, now + 0.08);
  gain.gain.linearRampToValueAtTime(0.22 * _uiSfxVolume, now + 1.0);
  gain.gain.linearRampToValueAtTime(0, now + 1.35);

  filter.frequency.setValueAtTime(120, now);
  filter.frequency.exponentialRampToValueAtTime(4800, now + 1.2);

  const snapSynth = new Tone.MetalSynth({
    envelope: { attack: 0.001, decay: 0.3, release: 0.1 },
    harmonicity: 5.1,
    modulationIndex: 32,
    resonance: 4000,
    octaves: 1.5
  }).toDestination();
  snapSynth.volume.value = -8 + volumeToDb(_uiSfxVolume);
  snapSynth.triggerAttackRelease("400", "8n", now + 1.4);

  window.setTimeout(() => {
    noise.dispose();
    filter.dispose();
    gain.dispose();
    snapSynth.dispose();
  }, 2500);
}

export async function playCountdownBeep() {
  if (!canPlayUiSfx()) return;
  await ensureStarted();
  const synth = new Tone.Synth({
    oscillator: { type: "sine" },
    envelope: { attack: 0.001, decay: 0.12, sustain: 0, release: 0.05 },
    volume: -6 + volumeToDb(_uiSfxVolume)
  }).toDestination();
  synth.triggerAttackRelease("A5", "16n");
  window.setTimeout(() => synth.dispose(), 500);
}

export async function playCountdownGo() {
  if (!canPlayUiSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "square" },
    envelope: { attack: 0.01, decay: 0.2, sustain: 0.1, release: 0.5 },
    volume: -3 + volumeToDb(_uiSfxVolume)
  });
  const now = Tone.now();
  ["A5", "D6", "G6"].forEach((note, index) => {
    synth.triggerAttackRelease(note, "8n", now + index * 0.1);
  });
  window.setTimeout(() => synth.dispose(), 2000);
}

export async function playRoundResult() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "sine" },
    envelope: { attack: 0.01, decay: 0.15, sustain: 0.05, release: 0.3 },
    volume: -10 + volumeToDb(_gameSfxVolume)
  });
  const now = Tone.now();
  synth.triggerAttackRelease("G5", "8n", now);
  synth.triggerAttackRelease("C6", "8n", now + 0.12);
  window.setTimeout(() => synth.dispose(), 1000);
}

export async function playWin() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "square" },
    envelope: { attack: 0.01, decay: 0.18, sustain: 0.1, release: 0.5 },
    volume: -4 + volumeToDb(_gameSfxVolume)
  });
  const now = Tone.now();
  ["C5", "E5", "G5", "C6", "E6", "G6"].forEach((note, index) => {
    synth.triggerAttackRelease(note, "8n", now + index * 0.1);
  });
  window.setTimeout(() => synth.dispose(), 2000);
}

export async function playLose() {
  if (!canPlayGameSfx()) return;
  await ensureStarted();
  const synth = new Tone.PolySynth(Tone.Synth).toDestination();
  synth.set({
    oscillator: { type: "sawtooth" },
    envelope: { attack: 0.02, decay: 0.2, sustain: 0.05, release: 0.4 },
    volume: -6 + volumeToDb(_gameSfxVolume)
  });
  const now = Tone.now();
  ["G4", "F4", "Eb4", "C4"].forEach((note, index) => {
    synth.triggerAttackRelease(note, "8n", now + index * 0.13);
  });
  window.setTimeout(() => synth.dispose(), 2000);
}

// -- Background music (procedural chiptune loop) ------------------------------

export async function startMenuMusic() {
  if (!canPlayMusic()) return;
  await ensureStarted();
  stopMusic();

  Tone.getTransport().bpm.value = 120;
  _musicSynth = new Tone.PolySynth(Tone.Synth).toDestination();
  _musicSynth.set({
    oscillator: { type: "square" },
    envelope: { attack: 0.02, decay: 0.1, sustain: 0.4, release: 0.8 },
    volume: -18 + volumeToDb(_musicVolume)
  });

  _musicReverb = new Tone.Reverb({ decay: 2, wet: 0.25 }).toDestination();
  _musicSynth.connect(_musicReverb);

  const pattern = [
    { time: "0:0", note: "C4", dur: "8n" },
    { time: "0:0:2", note: "E4", dur: "8n" },
    { time: "0:1", note: "G4", dur: "8n" },
    { time: "0:1:2", note: "C5", dur: "8n" },
    { time: "0:2", note: "E5", dur: "8n" },
    { time: "0:2:2", note: "G4", dur: "8n" },
    { time: "0:3", note: "E4", dur: "8n" },
    { time: "0:3:2", note: "C4", dur: "8n" },
    { time: "1:0", note: "D4", dur: "8n" },
    { time: "1:0:2", note: "F4", dur: "8n" },
    { time: "1:1", note: "A4", dur: "8n" },
    { time: "1:1:2", note: "D5", dur: "8n" },
    { time: "1:2", note: "F5", dur: "8n" },
    { time: "1:2:2", note: "A4", dur: "8n" },
    { time: "1:3", note: "F4", dur: "8n" },
    { time: "1:3:2", note: "D4", dur: "8n" }
  ];

  const synth = _musicSynth;
  if (!synth) return;

  _musicLoop = new Tone.Part((time, value: { note: string; dur: string }) => {
    synth.triggerAttackRelease(value.note, value.dur, time);
  }, pattern);
  _musicLoop.loop = true;
  _musicLoop.loopEnd = "2m";
  _musicLoop.start(0);
  Tone.getTransport().start();
}

export async function startMatchMusic() {
  if (!canPlayMusic()) return;
  await ensureStarted();
  stopMusic();

  Tone.getTransport().bpm.value = 140;
  _musicSynth = new Tone.PolySynth(Tone.Synth).toDestination();
  _musicSynth.set({
    oscillator: { type: "square" },
    envelope: { attack: 0.01, decay: 0.08, sustain: 0.3, release: 0.5 },
    volume: -20 + volumeToDb(_musicVolume)
  });

  const pattern = [
    { time: "0:0", note: "A3", dur: "8n" },
    { time: "0:0:2", note: "C4", dur: "8n" },
    { time: "0:1", note: "E4", dur: "8n" },
    { time: "0:1:2", note: "A4", dur: "8n" },
    { time: "0:2", note: "G4", dur: "8n" },
    { time: "0:2:2", note: "E4", dur: "8n" },
    { time: "0:3", note: "C4", dur: "8n" },
    { time: "0:3:2", note: "A3", dur: "8n" },
    { time: "1:0", note: "F3", dur: "8n" },
    { time: "1:0:2", note: "A3", dur: "8n" },
    { time: "1:1", note: "C4", dur: "8n" },
    { time: "1:1:2", note: "F4", dur: "8n" },
    { time: "1:2", note: "E4", dur: "8n" },
    { time: "1:2:2", note: "C4", dur: "8n" },
    { time: "1:3", note: "A3", dur: "8n" },
    { time: "1:3:2", note: "F3", dur: "8n" }
  ];

  const synth = _musicSynth;
  if (!synth) return;

  _musicLoop = new Tone.Part((time, value: { note: string; dur: string }) => {
    synth.triggerAttackRelease(value.note, value.dur, time);
  }, pattern);
  _musicLoop.loop = true;
  _musicLoop.loopEnd = "2m";
  _musicLoop.start(0);
  Tone.getTransport().start();
}

export function stopMusic() {
  _musicLoop?.stop();
  _musicLoop?.dispose();
  _musicLoop = null;

  _musicSynth?.dispose();
  _musicSynth = null;

  _musicReverb?.dispose();
  _musicReverb = null;

  Tone.getTransport().stop();
}
